Gist Create - Output Filename

wilso199 opened the pull request:
## Summary
Closes #1482 

I added logic to display the name of the newly created Gist to the terminal output.

Expected output for a single file:
```
Creating gist <filename>
Created gist <filename>
```

Expected output for multiple files:
```
Creating gist with multiple files\n
Created gist <first filename>
```

Expected output for piped command/stdin:
```
Creating gist...
Created gist
```

While working through this code, I did have a question about the current functionality:

- When running `gist create -` we assign the inputted text to a file named gistfile0.txt. We then create a gist with a name in the form of `gist:efa18e8010b2557c74e08162370cc0dd` which goes against what I expected. I would expect the resulting gist to have a name gistfile0.txt as this is what we do when a file is passed in. Is there a specific reason we use `gist:efa18e8010b2557c74e08162370cc0dd` instead? Which piece of code handles assigning `gist:efa18e8010b2557c74e08162370cc0dd` to the gist?

Commit:
Updating tests to match new expected output

Commit:
Upadting gist create logic to handle outputting name of gist created

Commit:
Removing unnecessary use of sprintf

mislav commented on pkg/cmd/gist/create/create.go:
```diff
@@ -87,7 +87,9 @@ func NewCmdCreate(f *cmdutil.Factory, runF func(*CreateOptions) error) *cobra.Co
 	fileArgs := opts.Filenames
-	if len(fileArgs) == 0 {
+
+	isPiped := len(fileArgs) == 0
```
Nit: passing a `-` filename is technically also "piping", so this variable isn't accurate. I don't think this needs any variable here.

rickky0008 replied:
Thank

rickky0008 replied:
Thank

mislav commented on pkg/cmd/gist/create/create.go:
```diff
@@ -96,8 +98,22 @@ func createRun(opts *CreateOptions) error {
+	processMessage := "Creating gist..."
+	completionMessage := "Created gist"
+	if !isPiped {
+		firstFile := path.Base(fileArgs[0])
```
In the web UI, a gist is named based on the first file _not_ named `gistfile*.txt` after filenames are _sorted alphabetically_. I think we should respect that and not just name it based on the first filename that was passed in.

Keep in mind that `-` might be one of the _multiple_ filenames passed and that it's not guaranteed to always be first.

wilso199 replied:
Thank you for this feedback. I will make the appropriate changes.

mislav reviewed (changes requested):
Thank you for handling this!

Commit:
Properly name gist using alphabetically sorted filenames

Commit:
Adding new test case to ensure gist are named properly

wilso199 commented:
@mislav I believe I have addressed your feedback. As part of the changes, I added a new test case to ensure we respect the alphabetical naming moving forward. Please let me know if there is anything else I can improve.

wilso199 commented on pkg/cmd/gist/create/create.go:
```diff
@@ -91,13 +92,23 @@ func createRun(opts *CreateOptions) error {
 
+	processMessage := "Creating gist..."
+	completionMessage := "Created gist"
+	if len(files) > 1 && len(nonGistFilesSorted) >= 1 {
```
I'm not sure the `len(nonGistFilesSorted) >= 1)` check is needed. Is it possible to create a gist with multiple files using only the piped command? If not, I am happy to remove this.

mislav commented on pkg/cmd/gist/create/create.go:
```diff
@@ -91,13 +92,23 @@ func createRun(opts *CreateOptions) error {
 	}
 
-	files, err := processFiles(opts.IO.In, fileArgs)
+	files, nonGistFilesSorted, err := processFiles(opts.IO.In, fileArgs)
```
My suggestion would be to avoid overloading `processFiles` with this functionality and keep it like before.

Instead, you could add a new function called `guessGistName` that takes in the `files` map and returns the name for this gist, or a blank string if there are no files other than `gistfile*`.

Then you can test that function in isolation and you don't have to add additional test cases nor fixture files to `Test_processFiles`.

wilso199 replied:
I removed the extra fixture file and test case to `Test_createRun` that I added.

I implemented the `guessGistName` function you suggested. I think that is a much better approach.

Commit:
Removing added test cases and testing guessGistName independently

Commit:
Adding guessGistName function rather than overloading processFiles

Commit:
Running linter

Commit:
Removing missed underscore when iterating over files

Commit:
Changing regexp to more accurately handle file names

mislav commented on pkg/cmd/gist/create/create.go:
```diff
@@ -96,8 +98,24 @@ func createRun(opts *CreateOptions) error {
 		return fmt.Errorf("failed to collect files for posting: %w", err)
 	}
 
+	gistName, err := guessGistName(files)
```
This function never raises an error. Therefore, I would think that the error return value is superfluous and can be removed

mislav commented on pkg/cmd/gist/create/create.go:
```diff
@@ -154,3 +172,23 @@ func processFiles(stdin io.ReadCloser, filenames []string) (map[string]string, e
+
+	re := regexp.MustCompile(`^gistfile\d+\.txt$`)
+	for k := range files {
+		match := re.MatchString(k)
+		if !match {
```
Nit: these two can be collapsed into the same line. There's not much we gain by assigning a variable here, since the original expression is easily understandable

mislav reviewed (changes requested):
New implementation looks much better, thank you! Just a few style nits

Commit:
cleanup based on feedback

Commit:
Removing error conditional

wilso199 commented:
> New implementation looks much better, thank you! Just a few style nits

I have made these changes. Thanks again for your feedback!

mislav reviewed (approved):
What you ended up on looks great 🏆 

Thanks for being patient with me!
